doc/rst/language/spec.rst

```diff
@@ -1680,6 +1705,105 @@ defined in :ref:`Multiple_Variable_Declarations`.
 Multiple variables can be grouped together using a tuple notation as
 described in :ref:`Variable_Declarations_in_a_Tuple`.
 
+.. _Split_Initialization:
+
+Split Initialization
+~~~~~~~~~~~~~~~~~~~~
+
+When the ``initialization-part`` is present, it forms the initialization
+expression. For local variables, if the ``initialization-part`` is
+omitted, the compiler will search forward in the function for the first
+assignment statement(s) setting that variable that occur before the
+variable is otherwise used. It will search only within block declarations
+``{ }`` and conditionals. These assignment statements are called
+applicable assignment statements. The variable will be initialized at the
+applicable assignment instead of being assigned to. This feature is
+called split initialization.
+
+   *Example (simple-split-init.chpl)*
+
+   The combination of statements ``const x;`` and ``x = 5;`` in the below
+   example are equivalent to the declaration ``const x = 5;``.
+
+   .. code-block:: chapel
+
+      proc main() {
+        const x;
+        x = 5;
+        writeln(x);
+      }
+
+   .. BLOCK-test-chapeloutput
+
+      5
+
+
+   *Example (no-split-init.chpl)*
+
+   In the following code, the variable ``x`` is used before it is
+   assigned to, and so split initialization cannot apply to that
+   variable.
+
+   .. code-block:: chapel
+
+      proc main() {
+        const x;
+        writeln(x);
+        x = 5;
+      }
+
+   .. BLOCK-test-chapeloutput
+
+      no-split-init.chpl:1: In function 'main':
+      no-split-init.chpl:2: error: 'x' cannot be default initialized
+      no-split-init.chpl:3: note: 'x' is before being initialized here
+
+   *Example (split-cond-blocks-init.chpl)*
+
+   Split initialization can find the applicable assignment statement
+   within a nested block or conditional. When conditionals are involved,
+   there might be multiple applicable assignment statements representing
+   different branches.
+
+   .. code-block:: chapel
+
+      config const option = false;
+      proc main() {
+        const x;
+        if option {
+          x = 6;
+        } else {
+          {
+            x = 4;
+          }
+        }
+        writeln(x);
+      }
+
+   .. BLOCK-test-chapeloutput
+
+      4
+
+Split initialization does not apply:
+
+ * to module-level variables, fields, config variables, or ``extern``
+   variables.
+ * to variables that are referred to in nested functions
```

Comment:
"nested functions" is an implementation restriction, which should not be in the spec.